File: typescript/src/code-smells/oop-abusers/refused-bequest.ts
class BaseController {
  start(): void {
    console.log('starting')
  }

  stop(): void {
    console.log('stopping')
  }

  reset(): void {
    console.log('resetting')
  }
}

export class ReadOnlyController extends BaseController {
  start(): void {
  }

  stop(): void {
  }
}

export function demoRefusedBequest(readonly: boolean): void {
  const controller: BaseController = readonly ? new ReadOnlyController() : new BaseController()
  controller.start()
  controller.stop()
}

File: csharp/src/code-smells/oop-abusers/RefusedBequest.cs
namespace RefactoringAvanzado.CodeSmells.OopAbusers;

public class BaseController
{
    public virtual void Start()
    {
        Console.WriteLine("starting");
    }

    public virtual void Stop()
    {
        Console.WriteLine("stopping");
    }

    public virtual void Reset()
    {
        Console.WriteLine("resetting");
    }
}

public class ReadOnlyController : BaseController
{
    public override void Start()
    {
    }

    public override void Stop()
    {
    }
}

public static class RefusedBequest
{
    public static void DemoRefusedBequest(bool readOnly)
    {
        BaseController controller = readOnly ? new ReadOnlyController() : new BaseController();
        controller.Start();
        controller.Stop();
    }
}
